pageserver/src/layered_repository/storage_layer.rs

```diff
@@ -0,0 +1,113 @@
+use crate::relish::RelishTag;
+use crate::repository::WALRecord;
+use crate::walredo::WalRedoManager;
+use crate::ZTimelineId;
+use anyhow::Result;
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+use zenith_utils::lsn::Lsn;
+
+pub static ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; 8192]);
+
+///
+/// Represents a version of a page at a specific LSN. The LSN is the key of the
+/// entry in the 'page_versions' hash, it is not duplicated here.
+///
+/// A page version can be stored as a full page image, or as WAL record that needs
+/// to be applied over the previous page version to reconstruct this version.
+///
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PageVersion {
+    /// an 8kb page image
+    pub page_image: Option<Bytes>,
+    /// WAL record to get from previous page version to this one.
+    pub record: Option<WALRecord>,
+}
```

Comment:
This must always be one image OR one record, yes? If so, I think this should be changed to:
```
enum PageVersion {
    Image(Bytes),
    Record(WALRecord),
}
```
This better expresses what we want to allow, because it's impossible to have `{`None, None`}` or `{ Some(...), Some(...)`.

Reply:
> This must always be one image OR one record, yes?

Surprisingly, no. When the WAL is received and processed and the wal decoder calls put_*() functions, you only have an image or a record on each PageVersion. However, when a get_page_at_lsn() request comes in and we have to perform WAL redo to reconstruct a page image, we also store the reconstructed image back in the PageVersion. The rocksdb implementation had reasons to do that because of the way the GC is implemented there, see comment in ObjectTimeLine.get_page_at_lsn`. The Layered implementation does the same, although it wouldn't have to.

That's a bit strange when the in-memory layer gets frozen. We will store both the image and the WAL record in the on-disk layer too, if we happened to both in memory.

This is not very well-thought out, we need a more coherent plan on what to materialize and when and which versions to keep as images and which as WAL records. But as things stand, PageVersion can hold both. I'll add a comment on that for now..